### PR TITLE
Stats: Check plans to show upsell notice only for sites with free plans

### PR DESCRIPTION
**Pull request by dognose24**
Related to #79880 

* Use `isSiteOnPaidPlan` to check if the site has any **_paid_** WPCOM ~or Jetpack~ plan.
* Show upsell notices for **_WPCOM_** sites only with free plans, e.g., `WordPress.com Free` ~and `Jetpack Free`~.
* Introduce the selector `hasLoadedSitePlansFromServer` to prepare the notice after plans are loaded.

## Testing Instructions

* Spin up the change with the Calypso Live link.

#### Jetpack site
* Navigate to Stats > `Traffic` page for a `Jetpack` site without any `Jetpack` plan or Stats products.
* Ensure the upsell notice shows as previously.
* Purchase `Jetpack Security` for the site.
* Ensure the upsell notice still shows.
* Purchase `Jetpack Complete` for the site.
* Ensure the upsell notice does not display.

#### WPCOM site
* Navigate to Stats > `Traffic` page with the feature flag for a `Simple` site without any `WPCOM` plan or Stats products: `/stats/day/{site-slug}?flags=stats/paid-wpcom-stats`.
* Ensure the upsell notice shows as previously.
* Purchase `WordPress.com Personal` for the site.
* Ensure the upsell notice does not display.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by kangzj** [comment]
Hi @dognose24, the unit tests are failing for the PR. Could you please take a look? Thanks

File: client/state/sites/selectors/has-site-product-jetpack-stats.ts [hunk 1]
```diff
@@ -11,13 +13,22 @@ const hasSiteProductJetpackStats = (
...
 		( product ) =>
 			productHasStats( camelOrSnakeSlug( product ), onlyPaid ) && product.expiryStatus !== 'expired'
 	);
+
 	const sitePlan = getSitePlan( state, siteId );
 	const siteHasStatsPlan = productHasStats( sitePlan?.product_slug as string, onlyPaid );
-	return siteHasStatsPlan || !! siteHasStatsProduct;
+
+	// Check whether sites have paid plans of WPCOM.
+	const siteHasPaidPlan = isSiteOnPaidPlan( state, siteId || 0 );
```
Comment (kangzj) [inline client/state/sites/selectors/has-site-product-jetpack-stats.ts]: As we talked on Slack, the logic here is some interim solution, so I'm thinking, could we move these out to the notices and not to change anything here? The reason is that the file could be just the same even after we bundled Stats into WPCOM plans.

File: client/my-sites/stats/stats-notices/index.tsx [hunk 2]
```diff
@@ -93,9 +95,11 @@ const NewStatsNotices = ( { siteId, isOdysseyStats, statsPurchaseSuccess }: Stat
 
 	const { isLoading, isError, data: serverNoticesVisibility } = useNoticesVisibilityQuery( siteId );
 
+	// TODO: Integrate checking purchases and plans loaded state into `hasSiteProductJetpackStatsPaid`.
 	const hasLoadedPurchases = useSelector( ( state ) => hasLoadedSitePurchasesFromServer( state ) );
+	const hasLoadedPlans = useSelector( ( state ) => hasLoadedSitePlansFromServer( state, siteId ) );
 
-	if ( ! hasLoadedPurchases || isLoading || isError ) {
+	if ( ! hasLoadedPurchases || ! hasLoadedPlans || isLoading || isError ) {
```
Comment (kangzj) [inline client/my-sites/stats/stats-notices/index.tsx:112]: I'm wondering how this works for Odyssey - will it load successfully?

Comment (dognose24) [reply client/my-sites/stats/stats-notices/index.tsx:112]: Nice catch! Fixed in https://github.com/Automattic/wp-calypso/pull/81514/commits/06fc681d4c7ae80e79d2bfcadeda6159744a0ace.

File: client/state/sites/selectors/has-site-product-jetpack-stats.ts (hunk 1 above)
Comment (dognose24) [reply client/state/sites/selectors/has-site-product-jetpack-stats.ts]: Moved in https://github.com/Automattic/wp-calypso/pull/81514/commits/2e5035da810bdc99362937f3af6b287ec1714941.

**Review by kangzj — APPROVED** [review]
It works well for both Calypso and Odyssey Stats, and both WPCOM and Jetpack sites 👍

**Pull request merged**